src/mozanalysis/sizing.py

```diff
@@ -0,0 +1,383 @@
+import attr
+
+from mozanalysis.bq import sanitize_table_name_for_bq
+from mozanalysis.utils import hash_ish
+from mozanalysis.experiment import TimeLimits
+
+
+@attr.s(frozen=False, slots=True)
+class HistoricalTarget:
+    """Query historical data.
+
+    The methods here query data in a way compatible with the following
+    principles, which are important for experiment analysis:
+
+    Example usage (in a colab notebook):
+
+        from google.colab import auth
+        auth.authenticate_user()
+        print('Authenticated')
+
+        from mozanalysis.sizing import HistoricalTarget
+        from mozanalysis.bq import BigQueryContext
+        from mozanalysis.metrics.desktop import active_hours, uri_count
+        from mozanalysis.segments.desktop import new_unique_profiles, \
+            new_or_resurrected_v3
+
+        bq_context = BigQueryContext(
+            dataset_id='mbowerman',  # e.g. mine's mbowerman
+            project_id='moz-fx-data-bq-data-science'  # this is the default anyway
+        )
+
+        ht = HistoricalTarget(
+            experiment_name='my_test_name',
+            start_date='2021-01-01',
+            num_dates_enrollment=8
+        )
+
+        # Run the query and get the results as a DataFrame
+        res = ht.get_single_window_data(
+            bq_context,
+            metric_list = [
+                active_hours,
+                uri_count
+            ],
+            last_date_full_data='2021-01-30',
+            analysis_start_days=0,
+            analysis_length_days=21,
+            target_list = [
+                new_unique_profiles,
+                new_or_resurrected_v3
+            ]
+        )
+
+    Args:
+        experiment_name (str): A simple name of this analysis, which is used
+            to create table names when saving results in BigQuery.
+        start_date (str): e.g. '2019-01-01'. First date for historical data to be
+            retrieved.
+        num_dates_enrollment (int, optional): Only include this many dates
+            of dummy enrollments, where clients that satisfy the target
+            criteria are added to the analysis's dataset.
+            If ``None`` then use the maximum number of dates
+            as determined by the metric's analysis window and
+            ``last_date_full_data``. Typically ``7n+1``, e.g. ``8``. The
+            factor '7' removes weekly seasonality, and the ``+1`` accounts
+            for the fact that enrollment typically starts a few hours
+            before UTC midnight.
+        analysis_length (int, optional): Number of days to include for analysis
+        targets_sql (str, optional): The SQL query that was executed to create
+            the targets
+            table; only defined after get_single_window_data method is called.
+        metrics_sql (str, optional): The SQL query that was executed to create the final
+            results table; only defined after get_single_wondow_data method is
+            called.
```

Comment:
Are these actually parameters that someone might pass in? Or are they just attributes that (sometimes) exist in the object? If it's the latter, what if we instead just made them private attributes and provided a getter method:
```python
def get_targets_sql(self):
    if not self._targets_sql:
        raise ValueError("Target SQL not available, must call `get_single_window_data` first") # or something
    return self._targets_sql
```